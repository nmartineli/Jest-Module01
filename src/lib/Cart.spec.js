import Cart from './Cart';

describe('Cart', () => {
  let cart;
  let product = {
    title: 'Adidas running shoes - women',
    price: 35388, // 353.88 | R$ 353,88 -> float nem sempre é exato dentro do JS
  };

  let product2 = {
    title: 'Adidas running shoes - men',
    price: 41872,
  };

  beforeEach(() => {
    cart = new Cart(); // garante que antes de cada teste seja criada uma nova e única instância do cart
  });

  describe('getTotal()', () => {
    it('should return 0 when getTotal is executed on a newly created instance of Cart', () => {
      expect(cart.getTotal().getAmount()).toEqual(0);
    });

    it('should multiply quantity and price and receive the total amount ', () => {
      const item = {
        product,
        quantity: 2, // 707765
      };

      cart.add(item);
      expect(cart.getTotal().getAmount()).toEqual(70776);
    });

    it('ensure no more than one product exists at a time', () => {
      cart.add({
        product,
        quantity: 2,
      });
      cart.add({
        product,
        quantity: 1,
      });

      expect(cart.getTotal().getAmount()).toEqual(35388);
    });

    it('should update total when a product gets included and then removed', () => {
      cart.add({
        product,
        quantity: 2,
      });

      cart.add({
        product: product2,
        quantity: 1,
      });

      cart.remove(product);

      expect(cart.getTotal().getAmount()).toEqual(41872);
    });
  });

  describe('checkout()', () => {
    it('should return an object with the total and the list os items', () => {
      cart.add({
        product,
        quantity: 2,
      });

      cart.add({
        product: product2,
        quantity: 3,
      });

      expect(cart.checkout()).toMatchSnapshot();
    });

    it('should reset the cart when checkout() is called', () => {
      cart.add({
        product: product2,
        quantity: 3,
      });

      cart.checkout();

      expect(cart.getTotal().getAmount()).toEqual(0);
    });
  });

  describe('summary()', () => {
    it('should return an object with the total and the list the items when summary is called', () => {
      cart.add({
        product,
        quantity: 2,
      });

      cart.add({
        product: product2,
        quantity: 3,
      });

      expect(cart.summary()).toMatchSnapshot();
      expect(cart.getTotal().getAmount()).toBeGreaterThan(0);
    });
  });
});
