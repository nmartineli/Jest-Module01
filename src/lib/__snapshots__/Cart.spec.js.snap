// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart checkout() should return an object with the total and the list os items 1`] = `
{
  "items": [
    {
      "product": {
        "price": 35388,
        "title": "Adidas running shoes - women",
      },
      "quantity": 2,
    },
    {
      "product": {
        "price": 41872,
        "title": "Adidas running shoes - men",
      },
      "quantity": 3,
    },
  ],
  "total": 196392,
}
`;

exports[`Cart summary() should return an object with the total and the list the items when summary is called 1`] = `
{
  "formatted": "R$1,963.92",
  "items": [
    {
      "product": {
        "price": 35388,
        "title": "Adidas running shoes - women",
      },
      "quantity": 2,
    },
    {
      "product": {
        "price": 41872,
        "title": "Adidas running shoes - men",
      },
      "quantity": 3,
    },
  ],
  "total": {
    "amount": 196392,
    "currency": "BRL",
    "precision": 2,
  },
}
`;
